torch/csrc/jit/passes/shape_analysis.cpp

```diff
@@ -201,10 +200,48 @@ class ShapePropagator {
         continue;
       }
     }
-
     return tensor_types;
   }
 
+  c10::ScalarType typeToScalarType(c10::TypePtr type) {
+    if (auto cast = type->cast<CompleteTensorType>()) {
+      return cast->scalarType();
+    } else if(auto cast = type->cast<DimensionedTensorType>()) {
+      return cast->scalarType();
+    } else if (auto cast = type->cast<FloatType>()) {
+      return c10::ScalarType::Float;
+    } else if (auto cast = type->cast<IntType>()) {
+      return c10::ScalarType::Long;
+    } else if (auto cast = type->cast<BoolType>()) {
+      return c10::ScalarType::Bool;
+    }
+    return c10::ScalarType::Undefined;
+  }
+
+  c10::ScalarType getScalarType(Node *node) {
+    c10::ScalarType ret = c10::ScalarType::Undefined;
+    auto& args = node->schema().arguments();
+    for (size_t i = 0 ; i < args.size() ; i++ ) { //auto arg : args ) {
+      auto arg = args[i];
+      auto dtype = typeToScalarType(arg.type());
+      auto inputDtype = typeToScalarType(node->inputs()[i]->type());
+      if (dtype != c10::ScalarType::Undefined) { //}->cast<DimensionedTensorType>()) {
+        if (ret == c10::ScalarType::Undefined ) {
+          ret = dtype;
+        } else {
+          ret = c10::promoteTypes(ret, dtype);
+        }
+      } else if (inputDtype != c10::ScalarType::Undefined) {
```

Comment:
I'm not very familiar with jit, but in the existing code when I added print statements, shape analysis rarely if ever seemed to be correctly finding any actual scalar types for me to try to promote. 
Typically the type of the arg.type() was TensorType or NumberType, not float/int/bool, and so couldn't be used.

Reply:
@eellison see here for explanation

Reply:
Not sure i understand the comment still.